[#45772] use direct upload tokens for direct upload

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/45772
- added feature flag for the direct upload implementation in the backend
  - upload with shares are now "legacy", enable it with the feature flag legacy_upload_preparation
- use upload toasts for storage upload
- use drop zone for upload

Kharonus committed:
[#45772] Use direct upload tokens as provided by the NC app

- https://community.openproject.org/work_packages/45772
- added feature flag for the direct upload implementation in the backend
  - upload with shares are now "legacy", enable it with the feature flag legacy_upload_preparation
- use upload toasts for storage upload
- use drop zone for upload

Kharonus committed:
[#45772] replace location string with location id

Kharonus committed:
[#45772] removed upload service for webdav based upload

Kharonus committed:
[#45772] finalizing direct upload code increment

- added error toast when dropping many files
- added unit tests for new query

Kharonus committed:
[#45772] fixed eslint and failing spec

apfohl commented on modules/storages/app/common/storages/peripherals/storage_interaction/storage_queries.rb:
```diff
@@ -52,16 +52,23 @@ def download_link_query
-            retry_proc: with_refreshed_token_proc,
-            finalize_url:
-          )
+          if OpenProject::FeatureDecisions.legacy_upload_preparation_active?
```
Maybe we should define a point of time in the future to remove the legacy upload method. A WP or so?

Kharonus replied:
I'd couple it to the successful usage of the real direct upload for any amount of time in production.

apfohl commented on frontend/src/app/core/file-upload/op-file-upload.service.ts:
```diff
@@ -76,7 +80,10 @@ export class OpenProjectFileUploadService {
     const { uploads, finished } = this.upload(url, files);
     const mapped = finished
-      .then((result:HalResource[]) => result.map((el:HalResource) => ({ response: el, uploadUrl: el.staticDownloadLocation.href }))) as Promise<{ response:HalResource, uploadUrl:string }[]>;
+      .then((result:HalResource[]) => result.map((el:HalResource) => ({
```
`el`? Is it `element`?

Kharonus replied:
renamed

apfohl commented on frontend/src/app/core/file-upload/op-file-upload.service.ts:
```diff
@@ -98,6 +105,7 @@ export class OpenProjectFileUploadService {
    * @param {string} url
    * @param {UploadFile} file
    * @param {string} method
+   * @param {'text'|'json'} responseType
```
I would suggest, adding the return type to the function and remove the whole comment.

Kharonus replied:
fixed

apfohl commented on frontend/src/app/core/state/file-links/file-links.service.ts:
```diff
@@ -102,19 +101,10 @@ export class FileLinksResourceService extends ResourceCollectionService<IFileLin
-        createdByName: file.createdByName,
-        lastModifiedByName: file.lastModifiedByName,
-      },
+      originData: { ...file },
```
👍

apfohl commented on frontend/src/app/shared/components/storages/storage/storage.component.ts:
```diff
@@ -127,6 +136,8 @@ export class StorageComponent extends UntilDestroyedMixin implements OnInit, OnD
       uploadFailed: (fileName:string):string => this.i18n.t('js.storages.file_links.upload_error', { fileName }),
       linkingAfterUploadFailed: (fileName:string, workPackageId:string):string =>
         this.i18n.t('js.storages.file_links.link_uploaded_file_error', { fileName, workPackageId }),
+      draggingManyFiles: (storageType:string):string => this.i18n.t('js.storages.file.dragging_many_files', { storageType }),
```
`draggingTooManyFiles`?

Kharonus replied:
I see no reason for making it longer. ;) `many` is quite the correct word if you want to emphasize "more the one".

apfohl commented on frontend/src/app/shared/components/storages/storage/storage.component.ts:
```diff
@@ -300,6 +312,58 @@ export class StorageComponent extends UntilDestroyedMixin implements OnInit, OnD
+    return observable
+      .pipe(
+        tap(() => {
+          setTimeout(() => this.toastService.remove(notification), 700);
```
Not for this PR, but wouldn't it be nice if the close timeout could be configured while creating the toast and kept track in the `ToastService`? And maybe not with `setTimeout` nut with `pipe(delay(...))`? This side effect here bothers me.

Kharonus replied:
yes, sounds sensible, yet the refactoring of the toast service would exceed the current scope of this WP.

apfohl commented on frontend/src/app/shared/components/toaster/upload-progress.component.ts:
```diff
@@ -48,85 +60,87 @@ import { UploadFile, UploadHttpEvent, UploadInProgress } from 'core-app/core/fil
+              break;
             default:
-            // Sent or unknown event
+            // unknown event
```
Maybe add a log message as well.

Kharonus replied:
added a  browser console warning

apfohl reviewed (changes requested):
Nothing seems to be wrong, but I still have some suggestions. Just ask me for another review, if you have changed something.

Kharonus committed:
[#45772] fix PR comments

- refactored method signature of `uploadSingle`
- log unknown event type from upload progress

Kharonus committed:
Merge branch 'dev' into implementation/45772-use-direct-upload-tokens-as-provided-by-the-nc-app

Kharonus committed:
[#45772] fixed avatar upload types

Kharonus committed:
[#45772] fixed direct upload for fog storage
